internal/netutil/netutil.go

```diff
@@ -0,0 +1,171 @@
+// for an HTTP request from one of various headers in order: Forwarded, X-Real-IP,
+// X-Forwarded-For. For the multi-valued Forwarded and X-Forwarded-For headers, the
+// first value in the list is returned. If the port is unknown, it will be zero.
+//
+// If the client is able to control the headers, they can control the result of this
+// function. The result should therefore not necessarily be trusted to be correct;
+// that depends on the presence and configuration of proxies in front of apm-server.
+func ClientAddrFromHeaders(header http.Header) (ip net.IP, port uint16) {
+	for _, parse := range parseHeadersInOrder {
+		if ip, port := parse(header); ip != nil {
+			return ip, port
+		}
+	}
+	return nil, 0
+}
+
+var parseHeadersInOrder = []func(http.Header) (net.IP, uint16){
+	parseForwardedHeader,
+	parseXRealIP,
+	parseXForwardedFor,
+}
+
+func parseForwardedHeader(header http.Header) (net.IP, uint16) {
+	forwarded := parseForwarded(getHeader(header, "Forwarded", "forwarded"))
+	if forwarded.For == "" {
+		return nil, 0
+	}
+	return ParseIPPort(MaybeSplitHostPort(forwarded.For))
+}
+
+func parseXRealIP(header http.Header) (net.IP, uint16) {
+	return ParseIPPort(MaybeSplitHostPort(getHeader(header, "X-Real-Ip", "x-real-ip")))
+}
+
+func parseXForwardedFor(header http.Header) (net.IP, uint16) {
+	if xff := getHeader(header, "X-Forwarded-For", "x-forwarded-for"); xff != "" {
+		if sep := strings.IndexRune(xff, ','); sep > 0 {
+			xff = xff[:sep]
+		}
+		return ParseIPPort(MaybeSplitHostPort(strings.TrimSpace(xff)))
+	}
+	return nil, 0
+}
+
+func getHeader(header http.Header, key, keyLower string) string {
+	if v := header.Get(key); v != "" {
+		return v
+	}
+
+	// header.Get() internally canonicalizes key names, but metadata.Pairs uses
+	// lowercase keys. Using the lowercase key name allows this function to be
+	// used for gRPC metadata.
+	if v, ok := header[keyLower]; ok && len(v) > 0 {
+		return v[0]
+	}
+	return ""
+}
+
+// forwardedHeader holds information extracted from a "Forwarded" HTTP header.
+type forwardedHeader struct {
+	For   string
+	Host  string
+	Proto string
+}
+
+// parseForwarded parses a "Forwarded" HTTP header.
+func parseForwarded(f string) forwardedHeader {
+	// We only consider the first value in the sequence,
+	// if there are multiple. Disregard everything after
+	// the first comma.
+	if comma := strings.IndexRune(f, ','); comma != -1 {
+		f = f[:comma]
+	}
+	var result forwardedHeader
+	for f != "" {
+		field := f
+		if semi := strings.IndexRune(f, ';'); semi != -1 {
+			field = f[:semi]
+			f = f[semi+1:]
+		} else {
+			f = ""
+		}
+		eq := strings.IndexRune(field, '=')
+		if eq == -1 {
+			// Malformed field, ignore.
+			continue
+		}
+		key := strings.TrimSpace(field[:eq])
+		value := strings.TrimSpace(field[eq+1:])
+		if len(value) > 0 && value[0] == '"' {
+			var err error
+			value, err = strconv.Unquote(value)
+			if err != nil {
+				// Malformed, ignore
+				continue
+			}
```

Comment:
just a semantic preference, but I find this easier to read:

```suggestion
			if val, err := strconv.Unquote(value); err == nil {
				value = val
			}
			// Malformed, ignore
```

Reply:
This code is just moved, so I'm going to leave this one. I don't think your suggestion is _quite_ right anyway, as "Malformed, ignore" is reached from the happy path.